CSVParser/README.md

```diff
@@ -0,0 +1,285 @@
+
+The `file` works similarly to `record` except that we use `crlf` as separator. 
+
+```lean
+def file : Parsec $ Array Record := manySep record (crlf <* notFollowedBy eof) <* (optional crlf) <* eof
+```
+
+The `notFollowedBy`, `optional` and `eof` is used to handle the optional new line at the end of the CSV file.
+
+### Parse it!
+
+Everything is done now. Let's parse a CSV file. We can now use the parser we wrote to parse the `csvString`. A `Parsec` takes a `String.Iterator` and returns the `ParseResult`.
+
+```lean
+#eval file csvString.mkIterator
+```
+
+If you are in VSCode you can move your cursor onto the `#eval` and you can see the result.
+
+![](./screenshot.png)
+
+The `ParseResult` return value is not very easy to cope with, so we can use a wrapper to lift it into an `Except`.
+
+```lean
+def parse (s : String) : Except String $ Array $ Array $ String :=
+  match file s.mkIterator with
+  | Parsec.ParseResult.success _ res => Except.ok res
+  | Parsec.ParseResult.error it err  => Except.error s!"offset {it.i.repr}: {err}"
+
+#eval parse csvString
+```
+
+## Some Extra Tricks
+
+### Code Structure
+
+To keep the code clear, the best practice is to split the data structure part and the parsing part into two modules, respectively `Basic.lean` and `Parser.lean`.
+
+So the file structure may look like:
+
+```sh
+.
+├── CSVParser.lean
+├── CSVParser
+│   ├── Basic.lean
+│   ├── Parser.lean
+│   └── Printer.lean
+```
+
+You can also have a separate `Printer` module if you have a complex serialization logic.
+
+It's also recommend group all functions in a separate namespace, namely `namespace CSVParser`.
+
+### User State
+
+If you have some experience of writing parser in Haskell, you will notice that in Lean `Parsec` we cannot have a user state, but that's a design on purpose. We should use the `StateT` transformer to encode the state.
+
+```lean
+abbrev StateParsec σ := StateT σ Parsec
+```
+
+And current `Parsec` function are automatically lifted from `Parsec` to `StateParsec`, so you can use them as usual.
+
+*NOTE:* the `abbrev` command is required, otherwise you will have to implement `MonadLift` yourself.
+
+### Chanllenge: How to parse CSV homogeneously?
+
+The parser we write above can parse heterogenous CSV files, which we may not want.
+
+```lean
+def csvString' : String := "a,\"b\nc\"\r\n1,2\r\n4,5,6"
+
+#eval parse csvString'
+-- Except.ok #[#["a", "b\nc"], #["1", "2"], #["4", "5", "6"]]
+```
+
+But how can we enforce that the lines are homogeneous?
+
+The answer of the question is provided in [Main.lean](./Main.lean), but please first think of it yourself. 
+
+*HINT:* Also similar to the `many*` series of combinators.
+
+### Further Thinking: How to use `Parsec` on `Stream`?
+
+Streaming is a very important feature for web. Protobuf, GIF, PNG, ..., all these formats are designed to support streaming, and we need a way to streamingly parse them in Lean.
```

Comment:
"for web" => "for the web or any application that wants to process data with minimal memory overhead"
"and we need a way to streamingly parse them in Lean" => "and so we should provide a parser in lean that can operate on streams"